### PR TITLE
Updated to core20

### LINKED ISSUE #53: Snap uses an old version of CMake
I have encountered an error where I have needed to add a dependency to my project that depends on CMake.

My Linux mint (20.2) installation includes CMake version `3.16.3`, yet in flutter the dependency in question sees a version of CMake `3.10.2`.

:question: Why is such an old version of CMake used in flutter-snap, while ubuntu's repos have updated to CMake `3.16.3`?

Related issue: https://github.com/objectbox/objectbox-dart/issues/317

---

For any developers currently running into this issue my advice is to follow the manual `tar.gz` installation process on the official flutter documentation.

https://docs.flutter.dev/get-started/install/linux#install-flutter-manually

### PR DESCRIPTION
**Pull request by kenvandine**
This branch updates to core20, still a WIP.  Fixes #53

### CODE REVIEW

File: env.sh [hunk 1]
```diff
@@ -9,13 +9,13 @@ SNAP_USER_DATA=$HOME/snap/flutter/current
...
 export GIT_CONFIG_NOSYSTEM=1
-export CURL_CA_BUNDLE=/snap/core18/current/etc/ssl/certs/ca-certificates.crt
-export GIT_SSL_CAINFO=/snap/core18/current/etc/ssl/certs/ca-certificates.crt
-export CPLUS_INCLUDE_PATH=$SNAP/usr/include/$SNAPCRAFT_ARCH_TRIPLET/c++/8:$SNAP/usr/include/c++/8:$SNAP/usr/include:$SNAP/usr/include/$SNAPCRAFT_ARCH_TRIPLET:$SNAP/usr/include/c++/8
-export LIBRARY_PATH=$SNAP/usr/lib/gcc/$SNAPCRAFT_ARCH_TRIPLET/8:$SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET:$SNAP/lib/$SNAPCRAFT_ARCH_TRIPLET:$SNAP/usr/lib
+export CURL_CA_BUNDLE=/snap/core20/current/etc/ssl/certs/ca-certificates.crt
+export GIT_SSL_CAINFO=/snap/core20/current/etc/ssl/certs/ca-certificates.crt
+export CPLUS_INCLUDE_PATH=$SNAP/usr/include/$SNAPCRAFT_ARCH_TRIPLET/c++/9:$SNAP/usr/include/c++/9:$SNAP/usr/include:$SNAP/usr/include/$SNAPCRAFT_ARCH_TRIPLET:$SNAP/usr/include/c++/9
+export LIBRARY_PATH=$SNAP/usr/lib/gcc/$SNAPCRAFT_ARCH_TRIPLET/9:$SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET:$SNAP/lib/$SNAPCRAFT_ARCH_TRIPLET:$SNAP/usr/lib
 export LDFLAGS="-lblkid -lgcrypt -llzma -llz4 -lgpg-error -luuid -lpthread -ldl $LDFLAGS"
```
Comment (jpnurmi) [inline env.sh:16]: does `-lepoxy` help?

**Comment by jpnurmi** [comment]
The analysis fails because `/lib/x86_64-linux-gnu/libgcc_s.so.1` is linked. I thought it was something implicitly pulled in by wrong `libepoxy.so` but that doesn't seem to be the case.

18.04: 
```
$ apt-file search /lib/x86_64-linux-gnu/libgcc_s.so.1
libgcc1: /lib/x86_64-linux-gnu/libgcc_s.so.1
```

20.04:
```
$ apt-file search /lib/x86_64-linux-gnu/libgcc_s.so.1
libgcc-s1: /lib/x86_64-linux-gnu/libgcc_s.so.1
```

We could try staging `libgcc-s1` next?

**Comment by ikbendewilliam** [comment]
Hi, thank you for your work! Do you have an ETA when this will be merged or is there still work to be done? I need this to build a snap that uses the [just_audio_libwinmedia](https://pub.dev/packages/just_audio_libwinmedia) plugin, see also [libwinmedia#8](https://github.com/harmonoid/libwinmedia/issues/8).
This is for a project for the [flutter hack challenge](https://flutterhack.devpost.com/) with 10 more days to deadline. 
Or can I install a snap from the working branch? (I'm not very familiar with snap).
If neither is possible it's fine, but I thought I'll at least ask 😁

**Comment by kenvandine** [comment]
> Hi, thank you for your work! Do you have an ETA when this will be merged or is there still work to be done? I need this to build a snap that uses the [just_audio_libwinmedia](https://pub.dev/packages/just_audio_libwinmedia) plugin, see also [libwinmedia#8](https://github.com/harmonoid/libwinmedia/issues/8). This is for a project for the [flutter hack challenge](https://flutterhack.devpost.com/) with 10 more days to deadline. Or can I install a snap from the working branch? (I'm not very familiar with snap). If neither is possible it's fine, but I thought I'll at least ask grin

@ikbendewilliam Here's an example of building a snap based on core20 using flutter.  [](https://gist.github.com/kenvandine/de8674a5eaf0d0c6c506bf38f91b9dcd)

I'd love to see what you come up with and am also happy to help you if you have questions while creating the snap.  Feel free to contact me!

**Comment by ikbendewilliam** [comment]
>@ikbendewilliam[](https://gist.github.com/kenvandine/de8674a5eaf0d0c6c506bf38f91b9dcd) Here's an example of building a snap based on core20 using flutter.

@kenvandine do you have a link perhaps, it seems to be missing? Thank you for your help!

This is the project I'm trying to build: [GitHub Repo](https://github.com/ikbendewilliam/flutter-puzzle-challenge), [Try it out on web](https://ikbendewilliam.github.io/flutter-puzzle-challenge/build/web/). Linux is the final platform that is not supported through a store.

(Microsoft rejected it because it's not a "game" according to them 😅 but that's another problem)

**Comment by kenvandine** [comment]
> > @ikbendewilliam Here's an example of building a snap based on core20 using flutter.
> 
> @kenvandine do you have a link perhaps, it seems to be missing? Thank you for your help!
> 
> This is the project I'm trying to build: [GitHub Repo](https://github.com/ikbendewilliam/flutter-puzzle-challenge), [Try it out on web](https://ikbendewilliam.github.io/flutter-puzzle-challenge/build/web/). Linux is the final platform that is not supported through a store.
> 
> (Microsoft rejected it because it's not a "game" according to them sweat_smile but that's another problem)

@ikbendewilliam sorry, mark down formatting issue.  I've edited my comment with the link

**Comment by Gustl22** [comment]
Any progress on this? Unfortunately I'm not familiar with creating my own snaps, and I used `snapcraft --use-lxd` option, which results in `error: unable to contact snap store`, and I do not have enough available space for multipass, see [this](https://forum.snapcraft.io/t/wy-multipass-has-hight-disk-usage/19840). 

Is there any reason to be still WIP? I would really appreciate if this could be released soon ;D

**Comment by zhangbanger** [comment]
Eagerly awaiting this as well :)

**Comment by chirag729** [comment]
@kenvandine Any progress on this issue?

**Comment by kenvandine** [comment]
I'm not sure when we're going to rebase on core20 or core22.  Flutter is intended to be available on 18.04 and there's a pretty easy way to build a snap based on core20 without the need for the flutter snap itself to be core20.  Have you run into issues building your snap based on core20 following the example I provided?

**Comment by zhangbanger** [comment]
@kenvandine I believe the issue is that the newest version of Flutter requires it in order to run flutter desktop for linux in #53 

See the [latest comment](https://github.com/canonical/flutter-snap/issues/53#issuecomment-1097241231)

**Comment by chirag729** [comment]
Yes I tried the workaround but ran into an error which I've been trying to solve without much luck. 
```
Failed to generate snap metadata: The specified command 'omnichess_app' defined in the app 'omnichess' does not exist.
```
I've debugged the snap and verified that omnichess_app exists so not sure why it's complaining.

I've had this working for a couple of months before I upgraded the audioplayers package to 1.0.0 which uses a function only available in cmake 3.11

**Comment by chirag729** [comment]
Another reason for updating is for local development. I can no longer use the snap to install flutter on my dev machine because the version of cmake it comes with is too old. Certain packages require higher versions so I'm forced to install manually which is not an ideal situation. I'd rather be using the exact same flutter install as my production apps as there are, strangely enough, subtle behavioural differences between the two

**Comment by kenvandine** [comment]
I've published the core20 build to the edge/core20 channel for further testing.  To test this:

snap refresh --channel=edge/core20 flutter

Any testing would be appreciated!

**Comment by zhangbanger** [comment]
I am now able to run Linux desktop on Flutter 3.0 right from the snap. Thanks!

**Comment by MateusAmin** [comment]
latest/edge/core20 is on core18 and has cmake 3.10.2

**Comment by kenvandine** [comment]
> 
I've reopened the core20 branch, you should get it automatically if you were tracking latest/edge/core20

@jpnurmi now that we've fixed the test infra, what do you think about merging this?

**Comment by MateusAmin** [comment]
@kenvandine

Thanks!

**Comment by jpnurmi** [comment]
> @jpnurmi now that we've fixed the test infra, what do you think about merging this?

I can't think of any reason why not. Do you have any particular concerns?

**Comment by kenvandine** [comment]
No, I think it's ready

**Comment by zhangbanger** [comment]
:partying_face:

**Comment by gsouf** [comment]
I tested core 20 as per above instructions. My app work like a charm, nothing to report

**Comment by zhangbanger** [comment]
Will this make it into the stable branch on snap? https://snapcraft.io/flutter looks like stable is still pointing to a commit in May

**Comment by kenvandine** [comment]
This has been promoted to stable.

**Pull request merged**